akka-docs/src/main/paradox/scala/fault-tolerance.md

```diff
@@ -31,27 +31,43 @@ in more depth.
 
 For the sake of demonstration let us consider the following strategy:
 
-@@snip [FaultHandlingDocSpec.scala]($code$/scala/docs/actor/FaultHandlingDocSpec.scala) { #strategy }
+Scala
+:  @@snip [FaultHandlingDocSpec.scala]($code$/scala/docs/actor/FaultHandlingDocSpec.scala) { #strategy }
 
-I have chosen a few well-known exception types in order to demonstrate the
+Java
+:  @@snip [FaultHandlingTest.java]($code$/java/jdocs/actor/FaultHandlingTest.java) { #strategy }
+
+We have chosen a few well-known exception types in order to demonstrate the
 application of the fault handling directives described in @ref:[supervision](general/supervision.md).
 First off, it is a one-for-one strategy, meaning that each child is treated
 separately (an all-for-one strategy works very similarly, the only difference
 is that any decision is applied to all children of the supervisor, not only the
-failing one). There are limits set on the restart frequency, namely maximum 10
-restarts per minute; each of these settings could be left out, which means
-that the respective limit does not apply, leaving the possibility to specify an
-absolute upper limit on the restarts or to make the restarts work infinitely.
-The child actor is stopped if the limit is exceeded.
+failing one). 
+In the above example, `10` and @scala[`1 minute`]@java[`Duration.create("1 minute")`] are passed to the `maxNrOfRetries`
+and `withinTimeRange` parameters respectively, which means that the strategy restarts a child up to 10 restarts per minute.
+The child actor is stopped if the restart count exceeds `maxNrOfRetries` during the `withinTimeRange` duration.
+
+Also, there are special values for these parameters. If you specify:
+
+* `-1` to `maxNrOfRetries`, and @scala[`Duration.inf`]@java[`Duration.Inf()`] to `withinTimeRange`
+    * then the child is always restarted without any limit
+* `-1` to `maxNrOfRetries`, and a non-infinite `Duration` to `withinTimeRange` 
+    * `maxNrOfRetries` is treated as `1`
+* a non-negative number to `maxNrOfRetries` and @scala[`Duration.inf`]@java[`Duration.Inf()`] to `withinTimeRange`
+    * `withinTimeRange` is treated as infinite duration (i.e.) no matter how long it takes, once the restart count exceeds `maxNrOfRetries`, the child actor is stopped  
```

Comment:
FYI, what I checked to confirm the above behavior.

[FaultHandling.scala#L238L242](https://github.com/akka/akka/blob/1d2748cad29c4b996a7ba96671e26dac963caa18/akka-actor/src/main/scala/akka/actor/FaultHandling.scala#L238L242)
```scala
  private[akka] def withinTimeRangeOption(withinTimeRange: Duration): Option[Duration] =
    if (withinTimeRange.isFinite && withinTimeRange >= Duration.Zero) Some(withinTimeRange) else None

  private[akka] def maxNrOfRetriesOption(maxNrOfRetries: Int): Option[Int] =
    if (maxNrOfRetries < 0) None else Some(maxNrOfRetries)
```

[FaultHandling.scala#L39L45](https://github.com/akka/akka/blob/1d2748cad29c4b996a7ba96671e26dac963caa18/akka-actor/src/main/scala/akka/actor/FaultHandling.scala#L39L45)  
`if (x.isDefined) x.get else 1` is what treats `maxNrOfRetries = -1` as `1`
```scala
  def requestRestartPermission(retriesWindow: (Option[Int], Option[Int])): Boolean =
    retriesWindow match {
      case (Some(retries), _) if retries < 1 ⇒ false
      case (Some(retries), None)             ⇒ { maxNrOfRetriesCount += 1; maxNrOfRetriesCount <= retries }
      case (x, Some(window))                 ⇒ retriesInWindowOkay(if (x.isDefined) x.get else 1, window)
      case (None, _)                         ⇒ true
    }
```

[FaultHandling.scala#L435](https://github.com/akka/akka/blob/1d2748cad29c4b996a7ba96671e26dac963caa18/akka-actor/src/main/scala/akka/actor/FaultHandling.scala#L435)
```scala
  private val retriesWindow = (
    SupervisorStrategy.maxNrOfRetriesOption(maxNrOfRetries),
    SupervisorStrategy.withinTimeRangeOption(withinTimeRange).map(_.toMillis.toInt))

    def processFailure(context: ActorContext, restart: Boolean, child: ActorRef, cause: Throwable, stats: ChildRestartStats, children: Iterable[ChildRestartStats]): Unit = {
    if (restart && stats.requestRestartPermission(retriesWindow))
      restartChild(child, cause, suspendFirst = false)
    else
      context.stop(child) //TODO optimization to drop child here already?
  }    
```